### ISSUE TITLE
Instancing and DotScene

### ISSUE THREAD

**Issue by paroj** [OGRECave/blender2ogre#67]
@paroj this has let me thinking about instancing and DotScene.

Do you have any ideas in that regard?

Perhaps DotScene could have a std::map that creates a new Instance Manager when there are more than 2 instances of a mesh.

Or perpahs it can be done with a group like the merge.<group>, would be something like instance.<group>.
And modify the DotScene DTD to have a tag regarding instancing with the names of the Instance Managers and the meshes.

_Originally posted by @sercero in https://github.com/OGRECave/blender2ogre/issues/66#issuecomment-791513287_

**Comment by paroj** [comment]
there is already the `entity.static` field in the DTD, which serves the same purpose as instancing - but currently it is ignored by the DotScenePlugin. Maybe some of the preceding DotScene implementations did support it.

Note, when it comes to simple foliage, batching approaches like pagedgeometry or the built-in static-geometry are superior to instancing.

For the above reason I am reluctant to silently create InstanceManagers behind the users back. Also the current implementation only accounts for the *first* sub-mesh.

Instead, we could built upon the existing concept and do something like `entity.instanced=static`, `entity.instanced=ManagerName`.

**Comment by sercero** [comment]
Oh, thanks for the info.

I assumed that instanced rendering was superior to static geometry just because is is more modern.

Great, I will look into that.

**Comment by sercero** [comment]
@paroj how do you think that the user could specify that some objects are instanced/static from the Blender side?

I was thinking of using the same solution as merging objects, that is creating groups like: `instanced.ManagerName` or `instanced.static`

In the first case we set `entity.instanced=ManagerName` and in the second case `entity.instanced=static`.

Do you think it is a good solution?

**Comment by paroj** [comment]
> I was thinking of using the same solution as merging objects, that is creating groups like: instanced.ManagerName or instanced.static

Sounds good to me. However, you would need to use "instanced.ManagerName" for both as the static geometry also uses named managers.

in the DotScenePlugin we would then try both options (instance manger and static geometry) and fall back to normal entities (while warning the user about it).

Note, that static geometry is also an instancing technique (software instancing), so this is not a misnomer.

**Comment by sercero** [comment]
I don't understand the part about trying both options and falling back, I think that it would be better to use then static.ManagerName and instanced.ManagerName because static geometry has been around long enough that people thinks that is it different to instanced.

Now that I think about it there is also the problem of specifying what type of instancing should be used, aren't there like 4 techniques?

**Comment by paroj** [comment]
> static.ManagerName and instanced.ManagerName

yes, being explicit what you want is probably better. Then we could also check the data at export time: e.g. static allows mixing different meshes, while instanced does not.

> Now that I think about it there is also the problem of specifying what type of instancing should be used, aren't there like 4 techniques?

yes, both managers have some parameters, that we cannot easily specify in blender. Therefore, I suggest that DotScenePlugin assumes that "ManagerName" was created beforehand.

However, in the case of instanced, the mesh currently needs to be specified at creation time. This is unfortunate, but I would start like that anyway.
We can then adapt the Ogre Instanced API to suit this. Otherwise we would have to create a duplicate XML node hierarchy, just for Instanced.

**Comment by sercero** [comment]
@paroj what do you think about changing the DTD like this:
```
<!ELEMENT entity (userData?)>
<!ATTLIST entity
	name		CDATA	#IMPLIED
	id		ID	#IMPLIED
	meshFile	CDATA	#REQUIRED
	material	CDATA	#IMPLIED

	<!-- static		(true | false) "false" -->
	<!-- Replace with: -->
	instancing	(none | static | hardware) "none"
	instanceManager	CDATA	#IMPLIED

	castShadows	(true | false) "true"
>
```

**Comment by paroj** [comment]
I think I would rather go with
```
static       CDATA  #IMPLIED
instanced    CDATA  #IMPLIED
```
which would mirror the UI in blender and allow keeping the DTD version as we would not be dropping the "static" field.

This would also avoid the inconsistent declaration like when one misses "instanceManager" while specifying "instancing != none".

**Comment by sercero** [comment]
Hey @paroj,

I was able to integrate Static Geometry without much trouble using your ideas.

The way it is now the user has to create the group beforehand (`mSceneMgr->createStaticGeometry("Group")`) and do the `sg->build()` after decoding the scene otherwise nothing shows.

We could even do the `createStaticGeometry()` and `build()` inside DotSceneLoader if we wanted and have the user do a `getStaticGeometry()` outside the plugin to change the parameters. All thanks to the simplicity of this API.
(I think this would be the best user experience)

Now, the *big* problem is with Instancing as you can imagine, because the user should do first: 
`mSceneMgr->createInstanceManager("Group", "Cube.mesh", "Scene", Ogre::InstanceManager::ShaderBased, 80, IM_USEALL, 0);`

So the user has to know the mesh name that will be used, the number of submeshes and which submesh should be instanced, not a great API.

On the other hand inside the DotScene Plugin I'm facing the problem of determining which material to use (`mSceneMgr->createInstancedEntity(materialName, managerName)`)

I was thinking that perhaps this could be done by loading the meshfile, going through the submeshes and determining the material name through that. But I don't understand why the `createInstancedEntity()` function cannot determine the material automatically... 
And it seems to be very inefficient as well.

What do you think?

**Comment by paroj** [comment]
> I was able to integrate Static Geometry without much trouble using your ideas.

:+1: 

> We could even do the createStaticGeometry() and build() inside DotSceneLoader if we wanted and have the user do a getStaticGeometry() outside the plugin to change the parameters. All thanks to the simplicity of this API.
(I think this would be the best user experience)

`build()` finalizes the StaticGeometry, which prevents changing some settings (marked with "Must be called before 'build'." in the docs).
Therefore, I would let the user call build(). To make clear that its the users responsibility, I would also let him call createStaticGeometry().

> Now, the big problem is with Instancing as you can imagine, because the user should do first:

I would start with this nevertheless. It somewhat resembles the StaticGeometry sequence sketched above and we can improve on that later; e.g. one could defer specifying the mesh until the first createInstancedEntity.
However, the core idea of instancing is that the same mesh is used for the whole batch.

Note, that you should keep to only supporting Meshes one submesh, because multiple submeshes will require stuff like:
https://ogrecave.github.io/ogre-next/api/2.1/instancing.html#InstancingMultipleSubmeshes

> On the other hand inside the DotScene Plugin I'm facing the problem of determining which material to use 

you could just get the original mesh from MeshManager and inspect that `->getSubMesh(0)->getMaterial()`

**Comment by paroj** [comment]
fixed in #115
